#191: add main logic for EidGeneratorStrategy

Chuburashka opened the pull request:
This is an approach for solving #191, i.e. allow different ways of generating the eid.

* add a database column `eid`.
* By default this will stay null.
* An EidGeneratorStrategy can be injected to generate this `eid` value in a different way (e.g. randomly).
* When sending events, the `eid` value is used for the events. If it's null (i.e. no strategy configured), we fall back to the current approach of converting the DB id into UID format.

**For the release notes:**

* #192: It's now possible to use a different way of generating the eid by providing an EidGeneratorStrategy.

Commit:
add main logic for EidGeneratorStrategy

Commit:
tests for EventLogWriter were fixed

ePaul commented on nakadi-producer-spring-boot-starter/src/test/java/org/zalando/nakadiproducer/config/EmbeddedDataSourceConfig.java:
```diff
@@ -23,4 +26,12 @@ public DataSource dataSource() throws IOException {
         return EmbeddedPostgres.start();
     }
+
+    @Bean
+    public TransactionTemplate requiresNewTransactionTemplate(
+        PlatformTransactionManager transactionManager) {
+        TransactionTemplate transactionTemplate = new TransactionTemplate(transactionManager);
+        transactionTemplate.setPropagationBehavior(TransactionDefinition.PROPAGATION_REQUIRES_NEW);
+        return transactionTemplate;
+    }
```
Why is this needed?

Chuburashka replied:
Removed
I forgot to delete it after debugging test with parallel transactions.

ePaul commented on nakadi-producer-spring-boot-starter/src/main/resources/db_nakadiproducer/migrations/V2/V2133546886.1.3__nakadi_event_eid.sql:
```diff
@@ -0,0 +1 @@
+ALTER TABLE nakadi_events.event_log ADD COLUMN eid uuid DEFAULT CAST(LPAD(TO_HEX(currval('nakadi_events.event_log_id_seq')), 32, '0') AS UUID);
```
What is the impact on this to entries which are already there when this is rolled out?
Do they stay NULL, or are they trying to fill the value from the sequence (which will break things, as during the migration, there is no `currval()` of the sequence).

Chuburashka replied:
They will be initialised with last value in `nakadi_events.event_log_id_seq`.
And it will be incorrect, because we will have the same value in the all these entries. 

I removed DEFAULT statement.

ePaul replied:
From my understanding, currval actually only works if the same transaction already did update the sequence. In the ALTER TABLE, that wouldn't be the case, so I think it would fail completely. But either way, removing the default expression is a good fix.

ePaul commented on nakadi-producer/src/main/java/org/zalando/nakadiproducer/transmission/impl/EventTransmissionService.java:
```diff
@@ -156,12 +156,16 @@ public void overrideClock(Clock clock) {
     /**
-     * Converts a number in UUID format.
+     * This is only needed for backward compatibility.
      *
      * <p>For instance 213 will be converted to "00000000-0000-0000-0000-0000000000d5"</p>
      */
-    private String convertToUUID(final int number) {
-        return new UUID(0, number).toString();
+    private String getEid(final EventLog event) {
+        if (event.getEid() == null) {
+            return new UUID(0, event.getId()).toString();
+        }
+
+        return event.getEid().toString();
     }
```
This could be an instance method on the EventLog class?

Chuburashka replied:
Moved to the EventLog class

ePaul commented on nakadi-producer/src/main/java/org/zalando/nakadiproducer/transmission/impl/EventTransmissionService.java:
```diff
@@ -156,12 +156,16 @@ public void overrideClock(Clock clock) {
-    private String convertToUUID(final int number) {
-        return new UUID(0, number).toString();
+    private String getEid(final EventLog event) {
+        if (event.getEid() == null) {
+            return new UUID(0, event.getId()).toString();
+        }
```
If we keep this, do we even need the complicated DEFAULT expression for the DB column?

Chuburashka replied:
Yes, I removed DEFAULT expression and this logic should work fine for backward compatibility.
First of all I removed DEFAULT, because it worked incorrectly with entries which are already in the table.

ePaul commented on nakadi-producer/src/main/java/org/zalando/nakadiproducer/eventlog/EidGeneratorStrategy.java:
```diff
@@ -0,0 +1,9 @@
+
+public interface EidGeneratorStrategy {
+
+  UUID generateEid();
```
I wonder whether we want to be more flexible here to allow for future extensions?
Like generating the UUID as a hash of the content?
Or a different one depending on the event type?

So my suggestion would be to have this method take some parameter(s) – either the EventLog object (though with the current approach this will be tricky, as it's only created afterwards) or the values that go into it (i.e. eventType, payload, compactionKey).

The two implementations we provide can just ignore these parameters, but if someone needs something specialized, we don't have to do another incompatible change.

Chuburashka replied:
I thought about that, I would prefer to use payload object for generation eid, but because we have just Object value I decided do nothing with that.

Added in last commit:
1. Moved `generateEid` call to the end [createEventLog method](https://github.com/Chuburashka/nakadi-producer-spring-boot-starter/blob/2eb8d38b676c15df70a736674ca4a9ba4a5962c9/nakadi-producer/src/main/java/org/zalando/nakadiproducer/eventlog/impl/EventLogMapper.java#L21)
2. Added EventLog as parameter.

Personally I not so much like this solution, because it a little bit tricky and depends on what time we initialise `eid` field.

Chuburashka replied:
Ok, I removed `EventLog`.
I agree that we can add additional method later.

For some extra cases I extracted `EventLogBuilder` interface and clients can implement this and build EventLog as they needed

ePaul commented on nakadi-producer-spring-boot-starter/src/main/java/org/zalando/nakadiproducer/eventlog/impl/eidgenerator/NoOpEidGeneratorStrategy.java:
```diff
@@ -0,0 +1,12 @@
+import java.util.UUID;
+import org.zalando.nakadiproducer.eventlog.EidGeneratorStrategy;
+
+public class NoOpEidGeneratorStrategy implements EidGeneratorStrategy {
```
Instead of public implementation classes (of this simple structure), I recently rather prefer static factory methods inside the interface.

I.e.
```java
public interface EidGeneratorStrategy {

    static EidGeneratorStrategy noop() {
        return () -> null;
    }

    static EidGeneratorStrategy random() {
        return () -> UUID.randomUUID();
    }

     UUID generateEid();
}
```

Chuburashka replied:
Thanks, did the same.

ePaul commented on nakadi-producer/src/main/java/org/zalando/nakadiproducer/eventlog/impl/EventLogMapper.java:
```diff
@@ -0,0 +1,41 @@
+import org.zalando.nakadiproducer.eventlog.EidGeneratorStrategy;
+import org.zalando.nakadiproducer.flowid.FlowIdComponent;
+
+public class EventLogMapper {
+
+  private final ObjectMapper objectMapper;
+  private final FlowIdComponent flowIdComponent;
+  private final EidGeneratorStrategy eidGeneratorStrategy;
```
Could you please follow the existing indentation style of the project (4 spaces)?

Chuburashka replied:
Fixed

Commit:
Removed DEFAULT value, changed logic generation for negative numbers
Tests, indentation were fixed.

Chuburashka commented on nakadi-producer-spring-boot-starter/src/main/resources/db_nakadiproducer/migrations/V2/V2133546886.1.3__nakadi_event_eid.sql:
```diff
@@ -0,0 +1 @@
+ALTER TABLE nakadi_events.event_log ADD COLUMN eid uuid DEFAULT CAST(LPAD(TO_HEX(currval('nakadi_events.event_log_id_seq')), 32, '0') AS UUID);
```
1. Fixed by adding this code: `CAST(CURRVAL('nakadi_events.event_log_id_seq') as BIGINT)`. 
We cast negative value to the BIGINT and after that apply `TO_HEX` function.
FYI: Sequence in the PostgreSQL with default configuration has a `MIN VALUE = 1` and by default it's impossible to has id less that 1.

2. fixed

ePaul replied:
Yes, negative values only happen if someone updates it manually, e.g. as a hot fix after the sequence runs out. (See #160 and the workaround mentioned there. We last week had one case where it was needed.)

ePaul commented on nakadi-producer/src/main/java/org/zalando/nakadiproducer/eventlog/EidGeneratorStrategy.java:
```diff
@@ -1,9 +1,22 @@
 
+    static EidGeneratorStrategy random() {
+        return (EventLog eventLog) -> UUID.randomUUID();
+    }
```
Let's add a bit of Javadocs to these methods:
```suggestion
    /**
     * A built-in strategy which does not generate an eid (which means the library will fall back
     * to converting the sequential DB ID into an UUID).
     * (This is the default strategy, and equivalent to the behavior before this interface was introduced.)
     */
    static EidGeneratorStrategy noop() {
        return (EventLog eventLog) -> null;
    }

    /**
     * A built-in strategy which will assign a random (type 4) UUID, ignoring the data.
     * You should only use this if your consumers don't depend on the eid for ordering
     * of events.
     */
    static EidGeneratorStrategy random() {
        return (EventLog eventLog) -> UUID.randomUUID();
    }
```

Chuburashka replied:
added

ePaul commented on nakadi-producer-spring-boot-starter/src/main/java/org/zalando/nakadiproducer/eventlog/impl/EventLogRepositoryImpl.java:
```diff
@@ -96,16 +96,18 @@ public void persist(Collection<EventLog> eventLogs) {
+              "    (event_type, event_body_data, flow_id, created, last_modified, locked_by, locked_until, compaction_key, eid)" +
               "VALUES " +
-              "    (:eventType, :eventBodyData, :flowId, :created, :lastModified, :lockedBy, :lockedUntil, :compactionKey)",
+              "    (:eventType, :eventBodyData, :flowId, :created, :lastModified, :lockedBy, :lockedUntil, :compactionKey, " +
+              "     COALESCE(:eid, CAST(LPAD(TO_HEX(CAST(CURRVAL('nakadi_events.event_log_id_seq') as BIGINT)), 32, '0') AS UUID)))",
```
If we keep the "convert on reading" fallback, then we can just put NULL here when using the noop strategy, right?
No need to do the complicated conversion on insert.
```suggestion
              "    (:eventType, :eventBodyData, :flowId, :created, :lastModified, :lockedBy, :lockedUntil, :compactionKey, " +
              "     :eid)",
```

Chuburashka replied:
I thought about that, but decided that if we provide a new column it would be better to provide a value too.
For me it's better to check data in the DB than to try to understand what the value will be send

But as you wish, in general I agree to remove this complex function.

ePaul replied:
You didn't change this, I think?

Chuburashka replied:
Yes, I left the function for now.
If you still think that better to remove it, I'll do it

Commit:
JavaDoc was added to EidGeneratorStrategy

Commit:
README was updated, some fixes by review

Commit:
EventLogBuilder interface was extracted

ePaul commented on nakadi-producer/src/main/java/org/zalando/nakadiproducer/eventlog/EventLogBuilder.java:
```diff
@@ -0,0 +1,8 @@
+
+import org.zalando.nakadiproducer.eventlog.impl.EventLog;
+
+public interface EventLogBuilder {
```
I think this should go into the impl package, as this is depending on the impl class EventLog, and should be something which normal library users don't need to be exposed to.
(Leaving it public is okay, maybe for corner cases.)

ePaul commented on nakadi-producer-spring-boot-starter/src/main/java/org/zalando/nakadiproducer/NakadiProducerAutoConfiguration.java:
```diff
@@ -186,8 +188,9 @@ public FlywayMigrator flywayMigrator() {
-                                         EidGeneratorStrategy eidGeneratorStrategy) {
-        return new EventLogMapper(objectMapper, flowIdComponent, eidGeneratorStrategy);
+    @ConditionalOnMissingBean
+    public EventLogBuilder eventLogMapper(ObjectMapper objectMapper, FlowIdComponent flowIdComponent,
```
```suggestion
    public EventLogBuilder eventLogBuilder(ObjectMapper objectMapper, FlowIdComponent flowIdComponent,
```

ePaul commented on README.md:
````diff
@@ -313,6 +313,15 @@ public SnapshotEventGenerator snapshotEventGenerator(MyService service) {
 }
 ```
 
````
We should also update the paragraph in line 21, e.g. something like this:

```diff
diff --git i/README.md w/README.md
index 491d563..a450741 100644
--- i/README.md
+++ w/README.md
@@ -18,7 +18,9 @@ There are already [multiple clients for the Nakadi REST API](https://zalando.git
 
 We solved them by persisting new events in a log table as part of the producing JDBC transaction. They will then be sent asynchronously to Nakadi after the transaction completed. If the transaction is rolled back, the events will vanish too. As a result, events will always be sent if and only if the transaction succeeded.
 
-The Transmitter generates a strictly monotonically increasing event id that can be used for ordering the events during retrieval. It is not guaranteed, that events will be sent to Nakadi in the order they have been produced. If an event could not be sent to Nakadi, the library will periodically retry the transmission.
+Unless a different [EID generation strategy](#eid-generation-strategy-optional) is used, the Transmitter generates a strictly monotonically increasing event id that can be used for ordering the events during retrieval.
+
+It is not guaranteed, that events will be sent to Nakadi in the order they have been produced. If an event could not be sent to Nakadi, the library will periodically retry the transmission.
 
 This project is mature, used in production in some services at Zalando, and in active development.
```

ePaul commented on nakadi-producer/src/main/java/org/zalando/nakadiproducer/eventlog/impl/EventLog.java:
```diff
@@ -26,4 +42,18 @@ public class EventLog {
+     * <p>For instance 213 will be converted to "00000000-0000-0000-0000-0000000000d5"</p>
+     */
+    public UUID getEid() {
+        if (eid == null) {
+            return new UUID(0, id);
+        }
+
+        return eid;
```
For performance reasons, we likely want to only do the conversion once.

```suggestion
        if (eid == null) {
            eid = new UUID(0, id);
        }

        return eid;
```

Currently we do it once before submitting, and once again after checking the failure results from Nakadi.
(It's been like that in the old code, but if we can improve this now, even better.)

This makes also the lookup a tiny bit faster, as it'll now be an == comparison before comparing the content.

ePaul commented on nakadi-producer/src/main/java/org/zalando/nakadiproducer/eventlog/impl/EventLog.java:
```diff
@@ -26,4 +42,18 @@ public class EventLog {
     private String compactionKey;
+    private UUID eid;
+
+    /**
+     * This is only needed for backward compatibility.
+     *
+     * <p>For instance 213 will be converted to "00000000-0000-0000-0000-0000000000d5"</p>
+     */
```
```suggestion
    /**
     * Returns the eid to be used for submitting the event. If none was stored, we'll convert it from the DB-ID.
     *
     * <p>For instance 213 will be converted to "00000000-0000-0000-0000-0000000000d5"</p>
     */
```

Commit:
JavaDoc's were updated, small refactoring

ePaul commented on nakadi-producer-spring-boot-starter/src/test/java/org/zalando/nakadiproducer/eventlog/impl/EventLogRepositoryIT.java:
```diff
@@ -77,8 +82,7 @@ public void testDeleteMultipleEvents() {
         assertThat(remaining, hasSize(1));
-        assertThat(remaining.get(0).getId(), is(notDeleted.getId()));
-        assertThat(remaining.get(0).getFlowId(), is(notDeleted.getFlowId()));
+        assertThat(remaining.get(0), equalTo(notDeleted));
```
I think this line worked when you wrote it, but now it fails, as EventLog doesn't have the generated equals method anymore, so it just does the `==` comparison.

I'd just revert this line of the change.

Chuburashka replied:
I changed `equalTo ` to `Matchers.samePropertyValuesAs`

Commit:
Tests fixed

Commit:
import was removed

Chuburashka commented on nakadi-producer/src/main/java/org/zalando/nakadiproducer/eventlog/impl/EventLog.java:
```diff
@@ -26,4 +42,19 @@ public class EventLog {
+        }
+
+        return eid;
+    }
```
Oh, sorry.
I ran `integration-test` task instead of `install`.

And `integration-test` executes succesfully and shown `BUILD SUCCESS` (although in the logs we can find NPE messages)

ePaul commented on nakadi-producer-spring-boot-starter/src/test/java/org/zalando/nakadiproducer/eventlog/impl/EventLogRepositoryIT.java:
```diff
@@ -77,8 +82,7 @@ public void testDeleteMultipleEvents() {
+        assertThat(remaining.get(0), Matchers.samePropertyValuesAs(notDeleted));
     }
 
     private List<EventLog> findAllEventsInDB() {
```
I'd change the query a few lines below to
```
"SELECT * FROM nakadi_events.event_log ORDER BY id ASC"
```
to get the events out in a reproducible order.

ePaul commented on nakadi-producer-spring-boot-starter/src/test/java/org/zalando/nakadiproducer/eventlog/impl/EventLogRepositoryIT.java:
```diff
@@ -88,18 +92,140 @@ private List<EventLog> findAllEventsInDB() {
 
+    @Test
+    @Transactional
+    public void testInsertEventWithDefaultEid() {
+        persistTestEvent("FLOW_ID");
+        EventLog actual = findAllEventsInDB().get(0);
+
+        EventLog expected = buildEventLog("FLOW_ID", 1, buildEid(1));
+        assertEvent(actual, expected);
+    }
```
I've expanded the test to make sure it's also working when inserting multiple events:
```suggestion
    public void testInsertSingleEventsWithDefaultEid() {
        persistTestEvent("FLOW_ID_1");
        persistTestEvent("FLOW_ID_2");

        List<EventLog> eventLogs = findAllEventsInDB();
        EventLog actual1 = eventLogs.get(0);
        EventLog actual2 = eventLogs.get(1);

        EventLog expected1 = buildEventLog("FLOW_ID_1", 1, buildEid(1));
        EventLog expected2 = buildEventLog("FLOW_ID_2", 2, buildEid(2));
        assertEvent(actual1, expected1);
        assertEvent(actual2, expected2);
    }

    @Test
    @Transactional
    public void testBulkInsertEventWithDefaultEid() {

        List<EventLog> eventLogsToPersist = Arrays.asList(
                buildEventLog("FLOW_ID_1"),
                buildEventLog("FLOW_ID_2")
        );
        eventLogRepository.persist(eventLogsToPersist);

        List<EventLog> eventLogsFound = findAllEventsInDB();
        EventLog actual1 = eventLogsFound.get(0);
        EventLog actual2 = eventLogsFound.get(1);

        EventLog expected1 = buildEventLog("FLOW_ID_1", 1, buildEid(1));
        EventLog expected2 = buildEventLog("FLOW_ID_2", 2, buildEid(2));
        assertEvent(actual1, expected1);
        assertEvent(actual2, expected2);
    }
```

Chuburashka replied:
Done

Commit:
Tests for bulk insert were added

Commit:
Default function for eid calculation was removed

Chuburashka commented:
@ePaul 
I removed function from INSERT query in the last commit.

Also I reverted [this changes](https://github.com/zalando-nakadi/nakadi-producer-spring-boot-starter/pull/192#discussion_r1649015096) for `EventLog.getEid()` method.

Because I think if we decided to leave `eid` field null by default better to leave it null and client side should decide what to do with this field. And we do it in the same place as before in [EventTransmissionService]( https://github.com/Chuburashka/nakadi-producer-spring-boot-starter/blob/1a49e50355a974f58aa1ab1a55ea4518434677d1/nakadi-producer/src/main/java/org/zalando/nakadiproducer/transmission/impl/EventTransmissionService.java#L163)

ePaul commented on nakadi-producer-spring-boot-starter/src/test/java/org/zalando/nakadiproducer/config/EmbeddedDataSourceConfig.java:
```diff
@@ -9,6 +9,9 @@
 import java.io.IOException;
 
 import javax.sql.DataSource;
+import org.springframework.transaction.PlatformTransactionManager;
+import org.springframework.transaction.TransactionDefinition;
+import org.springframework.transaction.support.TransactionTemplate;
```
These seem to be the only changes in this file? Is this still needed?

Chuburashka replied:
Removed

Commit:
Removed unused imports

ePaul commented:
I released a release-candidate version (21.1.0-RC-eidgenerator) [to our internal Nexus](https://maven.zalando.net/#browse/search/maven=attributes.maven2.artifactId%3Dnakadi-producer:NX.coreui.model.Component-1).

Feel free to try this out in your projects.

There are some merge conflicts with some of the other open PRs, so I couldn't easily create a release candidate containing the changes for all of them.

Chuburashka commented on nakadi-producer/src/main/java/org/zalando/nakadiproducer/eventlog/EidGeneratorStrategy.java:
```diff
@@ -0,0 +1,31 @@
+        return () -> UUID.randomUUID();
+    }
+
+    UUID generateEid();
```
@ePaul 
Are you agree with UUID as a returned type?

ePaul replied:
What would speak against it? (I might be forgetting some discussion we had 3 weeks ago, please remind me if there was something.)

Maybe we should add a `@Nullable` annotation and add Javadoc explaining what returning `null` means, though.

Chuburashka replied:
I think we haven't spoken about it.
I asked because in the `fahrschein` and `nakadi` libraries eid is just a string.
And we can expect some client who uses not UUID for eid.

ePaul replied:
Nakadi's API requires a string in UUID-format.
Passing a non-UUID string will break things, so making it an UUID here is fine. (We of course need to convert it to a string before submitting to Nakadi, or rather before passing it to Fahrschein, but that's just a `.toString()`.)

Chuburashka replied:
Got it, thanks
I checked the nakadi and they validate eid field format as UUID

ePaul commented:
:+1:

Chuburashka commented:
👍

Commit:
Merge branch 'master' into configurable-eid-generation-strategy

Commit:
Merge branch 'master' into configurable-eid-generation-strategy

ePaul commented on nakadi-producer/src/main/java/org/zalando/nakadiproducer/eventlog/impl/EventLogWriterImpl.java:
```diff
@@ -13,32 +13,25 @@
 
     public EventLogWriterImpl(EventLogRepository eventLogRepository,
-                              ObjectMapper objectMapper,
-                              FlowIdComponent flowIdComponent,
+                              EventLogBuilder eventLogBuilder,
                               List<CompactionKeyExtractor> keyExtractors) {
```
I wonder whether we should keep the old constructor (just creating an EventLogBuilder and calling the new one), in case someone is using nakadi-producer without the starter?

Though our SBOM data doesn't show anyone using Nakadi-producer without the spring-boot-starter, so that might not be a concern.

Commit:
Add test and a bit of javadoc to the EidGeneratorStrategy

Commit:
formatting, more README

Commit:
README: add example for strategy bean
